core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncSigningStage.java

```diff
@@ -48,9 +63,100 @@ public AsyncSigningStage(HttpClientDependencies dependencies) {
     @Override
     public CompletableFuture<SdkHttpFullRequest> execute(SdkHttpFullRequest request, RequestExecutionContext context)
             throws Exception {
+        // TODO: Add unit tests for SRA signing logic.
+        if (context.executionAttributes().getAttribute(SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME) != null) {
+            return sraSignRequest(request, context);
+        }
         return signRequest(request, context);
     }
 
+    private <T extends Identity> CompletableFuture<SdkHttpFullRequest> sraSignRequest(SdkHttpFullRequest request,
+                                                                                      RequestExecutionContext context) {
+        updateInterceptorContext(request, context.executionContext());
+
+        ExecutionAttributes executionAttributes = context.executionAttributes();
+        SelectedAuthScheme<T> selectedAuthScheme =
+            executionAttributes.getAttribute(SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME);
+
+        if (!shouldSign(selectedAuthScheme)) {
+            return CompletableFuture.completedFuture(request);
+        }
+
+        Pair<SdkHttpFullRequest, Duration> measuredSign = MetricUtils.measureDuration(
+            () -> {
+                AuthSchemeOption authSchemeOption = selectedAuthScheme.authSchemeOption();
+
+                // TODO: Identity resolution should move to before Endpoint resolution interceptor, to support accountId based
+                //  endpoints and also to logically separate out identity resolution as its own step.
+                ResolveIdentityRequest.Builder identityRequestBuilder = ResolveIdentityRequest.builder();
+                authSchemeOption.forEachIdentityProperty(identityRequestBuilder::putProperty);
+
+                IdentityProvider<T> identityProvider = selectedAuthScheme.identityProvider();
+                T identity = identityProvider.resolveIdentity(identityRequestBuilder.build()).join();
+
+                HttpSigner<T> signer = selectedAuthScheme.signer();
+
+                SdkHttpFullRequest result;
+                if (context.requestProvider() == null) {
```

Comment:
It's weird we use content stream providers in the async request pipeline. Not your problem, just feels weird we're not consistent about it.